database/src/tables.rs

```diff
@@ -12,6 +13,76 @@ use crate::{
     },
 };
 
+//---------------------------------------------------------------------------------------------------- Tables
+/// TODO
+macro_rules! define_trait_tables {
+    ($(
+        $table:ident => $index:literal
+    ),* $(,)?) => { paste::paste! {
+        /// TODO
+        #[allow(missing_docs)]
+        pub trait Tables {
+            $(
+                fn [<$table:snake>](&self) -> &(impl DatabaseRo<$table> + DatabaseIter<$table>);
+            )*
+        }
+
+        /// TODO
+        #[allow(missing_docs)]
+        pub trait TablesMut {
+            $(
+                fn [<$table:snake _mut>](&mut self) -> &mut impl DatabaseRw<$table>;
+            )*
+        }
+
+        impl<$([<$table:upper>]),*> Tables
+            for ($([<$table:upper>]),*)
+        where
+            $(
+                [<$table:upper>]: DatabaseRo<$table> + DatabaseIter<$table>,
+            )*
+        {
+            $(
+                fn [<$table:snake>](&self) -> &(impl DatabaseRo<$table> + DatabaseIter<$table>) {
+                    &self.$index
+                }
+            )*
+        }
+
+        impl<$([<$table:upper>]),*> TablesMut
+            for ($([<$table:upper>]),*)
+        where
+            $(
+                [<$table:upper>]: DatabaseRw<$table>,
+            )*
+        {
+            $(
+                fn [<$table:snake _mut>](&mut self) -> &mut impl DatabaseRw<$table> {
+                    &mut self.$index
+                }
+            )*
+        }
+    }};
+}
+
+define_trait_tables! {
+    BlockInfoV1s => 0,
+    BlockInfoV2s => 1,
+    BlockInfoV3s => 2,
+    BlockBlobs => 3,
+    BlockHeights => 4,
+    KeyImages => 5,
+    NumOutputs => 6,
+    PrunedTxBlobs => 7,
+    PrunableHashes => 8,
+    Outputs => 9,
+    PrunableTxBlobs => 10,
+    RctOutputs => 11,
+    TxIds => 12,
+    TxHeights => 13,
+    TxUnlockTime => 14,
+}
```

Comment:
This macro defines and blanket impls both `trait {Tables,TablesMut}`.

IMO I prefer manual impl for everything. For something that won't be changed often, the more verbose version is more clear in my opinion - type errors will still pop up when editing as well so it's still safe.

Reply:
It's not really a huge issue but I still think the macro is nicer, I don't think there is much difference in the clarity of the macro vs the verbose option, especially with docs.

Reply:
Fwiw, I also prefer the manual impl over the macro one.